webui/react/src/stores/cluster.tsx

```diff
@@ -0,0 +1,169 @@
+    Loadable.map(agents, (agents) => {
+      const overview: ClusterOverview = clone(initClusterOverview);
+
+      agents.forEach((agent) => {
+        agent.resources
+          .filter((resource) => resource.enabled)
+          .forEach((resource) => {
+            const isResourceFree = resource.container == null;
+            const availableResource = isResourceFree ? 1 : 0;
+            overview[resource.type].available += availableResource;
+            overview[resource.type].total++;
+            overview[ResourceType.ALL].available += availableResource;
+            overview[ResourceType.ALL].total++;
+          });
+      });
+
+      for (const key in overview) {
+        const rt = key as ResourceType;
+        overview[rt].allocation =
+          overview[rt].total !== 0
+            ? percent((overview[rt].total - overview[rt].available) / overview[rt].total)
+            : 0;
+      }
+
+      return overview;
+    }),
+  );
+
+  clusterStatus = Observable.select(
+    [this.clusterOverview, this.resourcePools, this.agents],
+    (overview, pools, agents) => {
+      return Loadable.match(Loadable.all([overview, pools, agents]), {
+        Loaded: ([overview, pools, agents]) => clusterStatusText(overview, pools, agents) ?? '',
+        NotLoaded: () => undefined,
+      });
+    },
+  );
+
+  fetchAgents = async (signal: AbortSignal): Promise<void> => {
+    try {
+      const response = await getAgents({}, { signal });
+      this.agents.set(Loaded(response));
+    } catch (e) {
+      handleError(e);
+    }
+  };
+
+  fetchResourcePools = async (signal: AbortSignal): Promise<void> => {
+    try {
+      const response = await getResourcePools({}, { signal });
+      this.resourcePools.set(Loaded(response));
+    } catch (e) {
+      handleError(e);
+    }
+  };
+
+  poll = (): void => {
+    this.canceler?.abort();
+    this.cancelPolling();
+    const canceler = new AbortController();
+    const pollFn = () => {
+      this.fetchResourcePools(canceler.signal);
+      this.fetchAgents(canceler.signal);
+    };
+
+    this.canceler = canceler;
+    pollFn();
+
+    this.pollingHandle = setInterval(pollFn, 5000);
+  };
+
+  cancelPolling = (): void => {
+    this.canceler?.abort();
+    if (!this.pollingHandle) return;
+    clearInterval(this.pollingHandle);
+  };
+}
+
+const StoreContext = createContext<ClusterService | null>(null);
+
+export const ClusterProvider = ({ children }: { children: ReactNode }): ReactElement => {
+  const [store] = useState(() => new ClusterService());
+
+  store.poll();
+
+  useEffect(() => store.cancelPolling, [store]);
+
+  return <StoreContext.Provider value={store}>{children}</StoreContext.Provider>;
+};
+
+const useClusterStore = (): ClusterService => {
+  const store = useContext(StoreContext);
+  if (store === null)
+    throw new Error('attempted to use cluster store outside of a cluster context');
+  return store;
+};
+
+export const useClusterStatus = (): string | undefined => {
```

Comment:
I'm not sold on having hooks for every store method, just seems like a lot of boilerplate to avoid calling `useObservable` in the relevant page. Just my opinion though.

Reply:
yeah only thing is that it makes it easier to switch observable libraries later, if we decide we want to do that. but that is not something i anticipate and i think that less boilerplate is good boilerplate.

Reply:
We're already re-exporting observables from our own file so I think that would be the level we'd most likely switch at.